File: pkg-py/tests/test_columns.py
"""Per-source synthetic columns exercise report/repair and the adapters uniformly.

The committed fixtures under _data/fixtures/columns give every source a mixed
column (valid / repairable / invalid / missing). These tests run the same column
through report(), repair(), is_valid_id, and the framework adapters, so all 46
sources get column-level coverage, not just mondo and hgnc.
"""

import json
from importlib.resources import files

import pytest

import biobouncer as bb

_COLUMNS = files("biobouncer") / "_data" / "fixtures" / "columns"
_SOURCES = sorted(bb.sources())
_CACHE_SOURCES = sorted(
    row["key"] for row in bb.source_info() if "cache" in row["modes"]
)


@pytest.fixture(autouse=True)
def _bundled_cache(tmp_path, monkeypatch):
    # Force cache mode to the bundled sample snapshots, as the fixtures were built.
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))


def _load(path):
    return [
        json.loads(line)
        for line in path.read_text(encoding="utf-8").splitlines()
        if line.strip()
    ]


def _rows(source_db):
    return _load(_COLUMNS / f"{source_db}.cases.jsonl")


def _cache_rows(source_db):
    return _load(_COLUMNS / f"{source_db}.cache.jsonl")


def _column(rows):
    return [row["input"] for row in rows]


@pytest.mark.parametrize("source_db", _SOURCES)
def test_report_summary_matches_categories(source_db):
    rows = _rows(source_db)
    categories = [row["category"] for row in rows]
    summary = bb.report(_column(rows), source_db, how="pattern").summary
    assert summary["total"] == len(rows)
    assert summary["valid"] == categories.count("valid")
    assert summary["repairable"] == categories.count("repairable")
    # invalid counts every failed value: the hard invalids plus the repairables.
    assert summary["invalid"] == categories.count("invalid") + categories.count(
        "repairable"
    )
    assert summary["missing"] == categories.count("missing")
    assert summary["indeterminate"] == 0


@pytest.mark.parametrize("source_db", _SOURCES)
def test_repair_substitutes_only_the_repairable_cells(source_db):
    rows = _rows(source_db)
    expected = [
        row["expect"]["suggestion"] if row["category"] == "repairable" else row["input"]
        for row in rows
    ]
    assert bb.report(_column(rows), source_db, how="pattern").repair() == expected


@pytest.mark.parametrize("source_db", _SOURCES)
def test_is_valid_id_per_row(source_db):
    rows = _rows(source_db)
    verdicts = bb.is_valid_id(_column(rows), source_db, how="pattern")
    for row, verdict in zip(rows, verdicts):
        if row["category"] == "valid":
            assert verdict is True
        elif row["category"] == "missing":
            assert verdict is None
        else:
            assert verdict is False


@pytest.mark.parametrize("source_db", _SOURCES)
def test_narwhals_mask_over_the_column(source_db):
    pytest.importorskip("narwhals")
    pd = pytest.importorskip("pandas")
    from biobouncer.narwhals import valid_id_mask

    rows = _rows(source_db)
    mask = list(
        valid_id_mask(pd.Series(_column(rows), name="id"), source_db, how="pattern")
    )
    for row, passes in zip(rows, mask):
        # Only an explicit failure masks False; a missing cell passes.
        assert passes == (row["category"] not in ("repairable", "invalid"))


@pytest.mark.parametrize("source_db", _SOURCES)
def test_pandera_flags_the_messy_column(source_db):
    pytest.importorskip("pandera")
    pd = pytest.importorskip("pandas")
    import pandera.pandas as pa

    from biobouncer.checks import is_id

    rows = _rows(source_db)
    schema = pa.DataFrameSchema(
        {"id": pa.Column(str, is_id(source_db, how="pattern"), nullable=True)}
    )
    # A valid-only column (with a missing cell) passes.
    valids = [row["input"] for row in rows if row["category"] == "valid"] + [None]
    schema.validate(pd.DataFrame({"id": valids}))
    # The full messy column always contains a hard invalid, so it must fail.
    with pytest.raises(pa.errors.SchemaError):
        schema.validate(pd.DataFrame({"id": _column(rows)}))


@pytest.mark.parametrize("source_db", _CACHE_SOURCES)
def test_cache_report_summary_matches_categories(source_db):
    rows = _cache_rows(source_db)
    categories = [row["category"] for row in rows]
    summary = bb.report(_column(rows), source_db, how="cache", version="sample").summary
    assert summary["total"] == len(rows)
    assert summary["valid"] == categories.count("valid")
    assert summary["repairable"] == categories.count("repairable")
    assert summary["invalid"] == categories.count("invalid") + categories.count(
        "repairable"
    )
    assert summary["missing"] == categories.count("missing")


@pytest.mark.parametrize("source_db", _CACHE_SOURCES)
def test_cache_repair_substitutes_only_the_repairable_cells(source_db):
    rows = _cache_rows(source_db)
    expected = [
        row["expect"]["suggestion"] if row["category"] == "repairable" else row["input"]
        for row in rows
    ]
    report = bb.report(_column(rows), source_db, how="cache", version="sample")
    assert report.repair() == expected


@pytest.mark.parametrize("source_db", _CACHE_SOURCES)
def test_cache_is_valid_id_per_row(source_db):
    rows = _cache_rows(source_db)
    verdicts = bb.is_valid_id(_column(rows), source_db, how="cache", version="sample")
    for row, verdict in zip(rows, verdicts):
        if row["category"] == "valid":
            assert verdict is True
        elif row["category"] == "missing":
            assert verdict is None
        else:
            assert verdict is False


@pytest.mark.parametrize("source_db", _CACHE_SOURCES)
def test_cache_narwhals_mask_over_the_column(source_db):
    pytest.importorskip("narwhals")
    pd = pytest.importorskip("pandas")
    from biobouncer.narwhals import valid_id_mask

    rows = _cache_rows(source_db)
    mask = list(
        valid_id_mask(
            pd.Series(_column(rows), name="id"),
            source_db,
            how="cache",
            version="sample",
        )
    )
    for row, passes in zip(rows, mask):
        assert passes == (row["category"] not in ("repairable", "invalid"))

File: pkg-r/tests/testthat/test-columns.R
# Per-source synthetic columns exercise report_id/repair_id and the adapters
# uniformly, so every source gets column-level coverage, not just mondo and hgnc.

.column_fixture <- function(source_db) {
  path <- file.path(
    system.file("extdata", "fixtures", "columns", package = "biobouncer"),
    paste0(source_db, ".cases.jsonl")
  )
  lines <- readLines(path, warn = FALSE)
  lines <- lines[nzchar(trimws(lines))]
  lapply(lines, jsonlite::fromJSON, simplifyVector = FALSE)
}

.fx_inputs <- function(fx) {
  vapply(
    fx,
    function(r) if (is.null(r$input)) NA_character_ else r$input,
    character(1)
  )
}

.fx_categories <- function(fx) {
  vapply(fx, function(r) r$category, character(1))
}

.fx_suggestions <- function(fx) {
  vapply(
    fx,
    function(r) {
      if (is.null(r$expect$suggestion)) NA_character_ else r$expect$suggestion
    },
    character(1)
  )
}

test_that("report_id summary matches the category tallies", {
  for (s in sources()) {
    fx <- .column_fixture(s)
    cats <- .fx_categories(fx)
    summ <- summary(report_id(.fx_inputs(fx), s, how = "pattern"))
    expect_equal(summ$total, length(fx))
    expect_equal(summ$valid, sum(cats == "valid"))
    expect_equal(summ$repairable, sum(cats == "repairable"))
    # invalid counts every failed value: the hard invalids plus the repairables.
    expect_equal(summ$invalid, sum(cats %in% c("invalid", "repairable")))
    expect_equal(summ$missing, sum(cats == "missing"))
    expect_equal(summ$indeterminate, 0)
  }
})

test_that("repair_id substitutes only the repairable cells", {
  for (s in sources()) {
    fx <- .column_fixture(s)
    cats <- .fx_categories(fx)
    inputs <- .fx_inputs(fx)
    expected <- ifelse(cats == "repairable", .fx_suggestions(fx), inputs)
    expect_identical(repair_id(inputs, s, how = "pattern"), expected)
  }
})

test_that("is_valid_id matches each cell category", {
  for (s in sources()) {
    fx <- .column_fixture(s)
    cats <- .fx_categories(fx)
    verdicts <- is_valid_id(.fx_inputs(fx), s, how = "pattern")
    expected <- ifelse(
      cats == "valid",
      TRUE,
      ifelse(cats == "missing", NA, FALSE)
    )
    expect_identical(verdicts, expected)
  }
})

test_that("the adapters flag repairable and invalid but pass a missing cell", {
  for (s in sources()) {
    fx <- .column_fixture(s)
    cats <- .fx_categories(fx)
    inputs <- .fx_inputs(fx)
    # id_predicate: TRUE for valid and missing (NA passes), FALSE for the rest.
    predicate <- id_predicate(s, how = "pattern")
    expect_identical(predicate(inputs), !(cats %in% c("repairable", "invalid")))
    # check_valid_id: a valid-only column with a missing cell passes; the messy
    # column always contains a hard invalid, so it returns a failure message.
    valids <- c(inputs[cats == "valid"], NA)
    expect_true(isTRUE(check_valid_id(valids, s, how = "pattern")))
    expect_type(check_valid_id(inputs, s, how = "pattern"), "character")
  }
})

.cache_col_sources <- function() {
  info <- source_info()
  info$key[grepl("cache", info$modes)]
}

.cache_column <- function(source_db) {
  path <- file.path(
    system.file("extdata", "fixtures", "columns", package = "biobouncer"),
    paste0(source_db, ".cache.jsonl")
  )
  lines <- readLines(path, warn = FALSE)
  lines <- lines[nzchar(trimws(lines))]
  lapply(lines, jsonlite::fromJSON, simplifyVector = FALSE)
}

test_that("report_id and repair_id work over a cache-mode column", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  for (s in .cache_col_sources()) {
    fx <- .cache_column(s)
    cats <- .fx_categories(fx)
    inputs <- .fx_inputs(fx)
    summ <- summary(report_id(inputs, s, how = "cache", version = "sample"))
    expect_equal(summ$valid, sum(cats == "valid"))
    expect_equal(summ$repairable, sum(cats == "repairable"))
    expect_equal(summ$invalid, sum(cats %in% c("invalid", "repairable")))
    expect_equal(summ$missing, sum(cats == "missing"))

    expected <- ifelse(cats == "repairable", .fx_suggestions(fx), inputs)
    expect_identical(
      repair_id(inputs, s, how = "cache", version = "sample"),
      expected
    )
  }
})

test_that("is_valid_id and the adapters work over a cache-mode column", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  for (s in .cache_col_sources()) {
    fx <- .cache_column(s)
    cats <- .fx_categories(fx)
    inputs <- .fx_inputs(fx)
    verdicts <- is_valid_id(inputs, s, how = "cache", version = "sample")
    expected <- ifelse(
      cats == "valid",
      TRUE,
      ifelse(cats == "missing", NA, FALSE)
    )
    expect_identical(verdicts, expected)

    predicate <- id_predicate(s, how = "cache", version = "sample")
    expect_identical(predicate(inputs), !(cats %in% c("repairable", "invalid")))
  }
})
